app/lib/differential_expression_service.rb

```diff
@@ -185,6 +153,93 @@ def self.run_differential_expression_job(cluster_file, study, user, annotation_n
     end
   end
 
+  # find all eligible annotations for DE for a given study
+  #
+  # * *params*
+  #   - +study+        (Study) => Associated study object
+  #   - +machine_type+ (String) => Override default VM machine type
+  #   - +dry_run+      (Boolean) => Indication of whether or not this is a pre-flight check
+  #   - +skip_existing+   (Boolean) => Skip annotations that already have DE results
+  #
+  # * *returns*
+  #   - (Array<Hash>) => Array of annotation objects available for DE
+  def self.find_eligible_annotations(study, machine_type: nil, dry_run: nil, skip_existing: false)
+    annotations = []
+    metadata = study.cell_metadata.where(annotation_type: 'group').select(&:can_visualize?)
+    annotations += metadata.map do |meta|
+      { annotation_name: meta.name, annotation_scope: 'study', machine_type: machine_type, dry_run: dry_run }
+    end
+    cell_annotations = []
+    groups_to_process = study.cluster_groups.select { |cg| cg.cell_annotations.any? }
+    groups_to_process.map do |cluster|
+      cell_annots = cluster.cell_annotations.select do |annot|
+        annot['type'] == 'group' && cluster.can_visualize_cell_annotation?(annot)
+      end
+      cell_annots.each do |annot|
+        annot[:cluster_file_id] = cluster.study_file.id # for checking associations later
+      end
+      cell_annotations += cell_annots
+    end
+    annotations += cell_annotations.map do |annot|
+      {
+        annotation_name: annot[:name],
+        annotation_scope: 'cluster',
+        cluster_file_id: annot[:cluster_file_id],
+        machine_type: machine_type,
+        dry_run: dry_run
```

Comment:
It's not a big deal in terms of maintainability, but I can't help but feel we should go beyond the trivial refactor I suggest below.  

It seems we're simply tacking on the same `machine_type` and `dry_run` to all of these annotation objects.  I suspect we could add them more simply / elegantly downstream of this method call, which would reduce visual noise and improve maintainability in this method.

```suggestion
        machine_type:,
        dry_run:
```

Reply:
See above - we do not allow mixed Hash literal styles in our linting.  I personally think this is a good rule, but I'm open to changing it if there's a salient case for it.  To me the mixed styles are hard to read and somewhat obtuse.

Reply:
See above for the linting comment.

As per refactoring: you're correct in that they're being passed all the way through because of the logic flow, and we need to get them at the final `:run_differential_expression_job`.   Let me take a pass at seeing if we can omit them until the very end to reduce the recapitulation throughout the stack.

Reply:
[d56bd23](https://github.com/broadinstitute/single_cell_portal_core/pull/1660/commits/d56bd23878ca4ea5107662706cfdc4991eed56c0)